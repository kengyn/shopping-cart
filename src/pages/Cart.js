import React, { useEffect } from "react";
import { useNavigate } from "react-router-dom";
import { items } from "../components/Items";
import "../styles/Cart.css";

const cart = [];

function addProduct(id, qty) {
  let products = items;
  let currentCart = cart;
  let product = products.find((item) => item.id === id);
  if (currentCart.some((item) => item.id.id === id)) {
    let itemIndex = currentCart.findIndex((i) => i.id.id === id);
    let updatedQty = parseInt(currentCart[itemIndex].qty) + parseInt(qty);
    let updatedItem = [{ id: product, qty: updatedQty }];

    let newCart = currentCart.map(
      (obj) => updatedItem.find((o) => o.id === obj.id) || obj
    );

    cart.splice(0, cart.length, ...newCart);
  } else {
    let bag = { id: product, qty };
    cart.push(bag);
  }
}

function totalQty() {
  let sum = cart.reduce((acc, obj) => {
    return acc + obj.qty;
  }, 0);
  return sum;
}

const Cart = () => {
  let navigate = useNavigate();
  const renderCart = () => {
    if (cart.length === 0) {
      return (
        <main>
          <img
            src={require("../assets/angryMayreel.gif")}
            alt="angry mayreel"
          />
          <div className="mayreel-cart-text">Cart Empty</div>
        </main>
      );
    }
    return (
      <>
        <h1>Cart</h1>
        <ul className="cart-list">
          {cart.map((item) => (
            <li key={item.id.id}>
              <img src={item.id.image} alt={item.id.name} />
              <div className="cart-product-name">{item.id.name}</div>
              <div className="cart-price-container">
                <div className="cart-price-info">
                  <div>Item Price</div>
                  <div>${item.id.price}.00</div>
                </div>
                <div className="cart-price-info">
                  <div>Quantity</div>
                  <div>{item.qty}</div>
                </div>

                <div className="cart-price-info">
                  <div>Total Price</div>
                  <div>${item.qty}.00</div>
                </div>
              </div>
              <div className="cart-buttons-container">
                <button id={item.id.id} className="cart-remove-button">
                  Remove
                </button>
              </div>
            </li>
          ))}
        </ul>
        <div className="checkout-container">
          <div className="checkout-title">Order Summary</div>
          <div className="sub-check-container">
            <div>Subtotal</div>
            <div>all ur money</div>
          </div>
          <div className="sub-check-container">
            <div>Tax</div>
            <div>gimme ur wallet</div>
          </div>
          <div className="sub-check-container estimate-container">
            <div>Estimated Total</div>
            <div>USD ${totalQty()}</div>
          </div>
          <button className="checkout-button">Checkout</button>
        </div>
      </>
    );
  };

  useEffect(() => {
    const removeButtons = document.querySelectorAll(".cart-remove-button");

    function deleteProduct(id) {
      let index = cart.findIndex((item) => item.id.id === id);
      if (index > -1) {
        cart.splice(index, 1);
      }
    }
    if (removeButtons) {
      removeButtons.forEach((button) => {
        button.addEventListener("click", () => {
          deleteProduct(button.id);
          navigate("/cart");
          console.log(cart);
        });
      });
    }
    return () => {
      if (removeButtons) {
        removeButtons.forEach((button) => {
          button.removeEventListener("click", () => {
            deleteProduct(button.id);
          });
        });
      }
    };
  }, [navigate]);

  return <main className="cart-main">{renderCart()}</main>;
};

export { addProduct, Cart, totalQty };
