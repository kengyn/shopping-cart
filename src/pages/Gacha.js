import React from "react";
import "../styles/Gacha.css";

export default function Gacha() {
  return (
    <main className="gacha-main">
      <img src={require("../assets/mayreel.gif")} alt="mayreel"></img>
      <div>Coming Soon</div>
    </main>
  );
}
