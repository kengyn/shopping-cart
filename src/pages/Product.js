import React, { useEffect } from "react";
import { useLoaderData, useNavigate } from "react-router-dom";
import { getProduct } from "../components/Items";
import "../styles/Product.css";

async function loader({ params }) {
  return getProduct(params.id);
}

const Product = () => {
  let navigate = useNavigate();
  const product = useLoaderData();
  let desc;

  if (product.id === "gem") {
    desc = (
      <img
        className="mayreel"
        src={require(`../assets/Mayreel.png`)}
        alt="mayreel"
      />
    );
  } else {
    desc = product.description;
  }

  useEffect(() => {
    const closeButton = document.querySelector(".close-button");
    const closeDisplay = document.querySelector(".display-product");

    function close() {
      navigate("/shop");
    }

    closeDisplay.addEventListener("click", (e) => {
      if (!e.target.classList.contains("display-product")) {
        return;
      } else {
        close();
      }
    });
    closeButton.addEventListener("click", () => close());

    return () => {
      closeDisplay.removeEventListener("click", () => close());
      closeButton.removeEventListener("click", () => close());
    };
  }, [navigate]);

  return (
    <div className="display-product">
      <div className="product-container">
        <img src={require(`../assets/${product.imgLink}`)} alt={product.name} />
        <div className="product-info">
          <div className="product-header">
            <div className="close-button">x</div>
            <div className="product-name">{product.name}</div>
            <div className="product-price">${product.price}</div>
          </div>
          <div className="product-description">
            <div>
              <div className="product-description-header">Game</div>
              <div>{product.game}</div>
            </div>
            <div>
              <div className="product-description-header">Banner</div>
              <div>{product.banner}</div>
            </div>
            <div>
              <div className="product-description-header">Description</div>
              <div>{desc}</div>
            </div>
          </div>
        </div>
        <div className="buy-container">
          <label htmlFor="quantity">QTY</label>
          <input type="number" id="quantity" defaultValue={1}></input>
          <button>Add To Cart</button>
        </div>
      </div>
    </div>
  );
};

export { loader, Product };
