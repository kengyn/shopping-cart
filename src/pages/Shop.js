import React from "react";
import { NavLink, Outlet } from "react-router-dom";
import { items } from "../components/Items";
import "../styles/Shop.css";

const Shop = () => {
  const renderItems = () => {
    return items.map((item) => (
      <li key={item.id} className={item.id}>
        <NavLink to={`product/${item.id}`}>
          <img src={item.image} alt={item.name} />
        </NavLink>
        <div>{item.name}</div>
        <div>${item.price}</div>
      </li>
    ));
  };

  return (
    <div className="shop-main">
      <h1>Shop</h1>
      <div className="items-container">
        <ul>{renderItems()}</ul>
      </div>
      <Outlet />
    </div>
  );
};

export default Shop;
