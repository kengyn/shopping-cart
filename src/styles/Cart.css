.mayreel-cart-text {
  font-weight: bold;
  font-size: 5rem;
}

.cart-main {
  display: grid;
  grid-template-areas:
    "title ."
    "item summary";
  column-gap: 2rem;
}

.cart-main h1 {
  grid-area: title;
  font-size: 1.6rem;
  margin-top: 1rem;
}

.cart-list {
  grid-area: item;
  display: flex;
  flex-direction: column;
}

.cart-list li {
  display: grid;
  grid-template-areas:
    "img name"
    "img price"
    "img remove";
  border-bottom: 1px solid lightgray;
  padding-top: 1rem;
}

.cart-list li img {
  grid-area: img;
  width: 150px;
  margin: 0;
}

.cart-product-name {
  grid-area: name;
  font-size: 1.3rem;
  font-weight: bold;
  align-self: flex-end;
  margin-bottom: 0.5rem;
}

.cart-price-container {
  grid-area: price;
  display: flex;
  gap: 1rem;
}

.cart-price-info {
  display: flex;
  flex-direction: column;
  align-items: flex-end;
}

.cart-buttons-container {
  grid-area: remove;
  display: flex;
  justify-content: flex-end;
}

.cart-remove-button {
  cursor: pointer;
  background-color: transparent;
  border: 1px solid transparent;
  border-bottom: 1px solid black;
  margin-bottom: 1rem;
}

.cart-remove-button:hover {
  border-bottom: 1px solid red;
}

.checkout-container {
  grid-area: summary;
  align-self: flex-start;
  justify-self: flex-end;
}

.checkout-title {
  font-size: 1.6rem;
  font-weight: bold;
  margin-bottom: 1rem;
}

.sub-check-container {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  padding-bottom: 1rem;
  margin-bottom: 1rem;
  border-bottom: 1px solid lightgray;
  gap: 4rem;
}

.estimate-container {
  border: none;
  font-size: 1.2rem;
  font-weight: bold;
}

.checkout-button {
  cursor: pointer;
  background-color: rgb(0 120 46);
  color: white;
  font-weight: bold;
  font-size: 1rem;
  width: 100%;
  height: 3.5rem;
  outline: none;
  border: none;
  transition: 0.3s;
}

.checkout-button:hover {
  background-color: #008533;
}
