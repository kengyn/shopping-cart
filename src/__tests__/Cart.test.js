import React from "react";
import { render } from "@testing-library/react";
import { Cart, addProduct } from "../pages/Cart";
import { MemoryRouter as Router } from "react-router-dom";

describe("cart component", () => {
  it("renders empty cart", () => {
    const { container } = render(
      <Router>
        <Cart />
      </Router>
    );

    expect(container).toMatchSnapshot();
  });

  it("renders item in cart correctly", () => {
    addProduct("acquaint-fate", 4);
    const { container } = render(
      <Router>
        <Cart />
      </Router>
    );
    expect(container).toMatchSnapshot();
  });
});
