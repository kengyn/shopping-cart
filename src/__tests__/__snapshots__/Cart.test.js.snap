// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cart component renders empty cart 1`] = `
<div>
  <main
    class="cart-main"
  >
    <main>
      <img
        alt="angry mayreel"
        src="angryMayreel.gif"
      />
      <div
        class="mayreel-cart-text"
      >
        Cart Empty
      </div>
    </main>
  </main>
</div>
`;

exports[`cart component renders item in cart correctly 1`] = `
<div>
  <main
    class="cart-main"
  >
    <h1>
      Cart
    </h1>
    <ul
      class="cart-list"
    >
      <li>
        <img
          alt="Acquaint Fate"
          src="Item_Acquaint_Fate.webp"
        />
        <div
          class="cart-product-name"
        >
          Acquaint Fate
        </div>
        <div
          class="cart-price-container"
        >
          <div
            class="cart-price-info"
          >
            <div>
              Item Price
            </div>
            <div>
              $
              1
              .00
            </div>
          </div>
          <div
            class="cart-price-info"
          >
            <div>
              Quantity
            </div>
            <div>
              4
            </div>
          </div>
          <div
            class="cart-price-info"
          >
            <div>
              Total Price
            </div>
            <div>
              $
              4
              .00
            </div>
          </div>
        </div>
        <div
          class="cart-buttons-container"
        >
          <button
            class="cart-remove-button"
            id="acquaint-fate"
          >
            Remove
          </button>
        </div>
      </li>
    </ul>
    <div
      class="checkout-container"
    >
      <div
        class="checkout-title"
      >
        Order Summary
      </div>
      <div
        class="sub-check-container"
      >
        <div>
          Subtotal
        </div>
        <div>
          all ur money
        </div>
      </div>
      <div
        class="sub-check-container"
      >
        <div>
          Tax
        </div>
        <div>
          gimme ur wallet
        </div>
      </div>
      <div
        class="sub-check-container estimate-container"
      >
        <div>
          Estimated Total
        </div>
        <div>
          USD $
          4
        </div>
      </div>
      <button
        class="checkout-button"
      >
        Checkout
      </button>
    </div>
  </main>
</div>
`;
